src/fifelse.c

```diff
@@ -140,3 +140,198 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
   UNPROTECT(nprotect);
   return ans;
 }
+
+SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
+  int n=length(args);
+  if (n % 2) {
+    error("Please supply an even number of arguments in ..., consisting of logical condition,"
+             " resulting value pairs (in that order); received %d inputs.", n);
+  }
+  int nprotect = 0, l = 0;
+  int64_t len0=0, len1=0, len2=0, idx=0;
+  SEXP ans = R_NilValue, value0 = R_NilValue, tracker = R_NilValue, cons = R_NilValue, outs = R_NilValue;
+  PROTECT_INDEX Icons, Iouts;
+  PROTECT_WITH_INDEX(cons, &Icons); nprotect++;
+  PROTECT_WITH_INDEX(outs, &Iouts); nprotect++;
+  SEXPTYPE type0;
+  bool nonna = !isNull(na), imask = true;
+  int *restrict p = NULL;
+  n = n/2;
+  for (int i=0; i<n; ++i) {
+    REPROTECT(cons = eval(VECTOR_PTR(args)[2*i], rho), Icons);
+    REPROTECT(outs = eval(VECTOR_PTR(args)[2*i+1], rho), Iouts);
+    if (!isLogical(cons)) {
+      error("Argument #%d must be logical.", 2*i+1);
+    }
+    const int *restrict pcons = LOGICAL(cons);
+    if (i == 0) {
+      len0 = xlength(cons);
+      len2 = len0;
+      type0 = TYPEOF(outs);
+      value0 = outs;
+      if (nonna) {
+        if (xlength(na) != 1) {
+          error("Length of 'default' must be 1.");
+        }
+        SEXPTYPE tn = TYPEOF(na);
+        if (tn == LGLSXP && LOGICAL(na)[0]==NA_LOGICAL) {
+          nonna = false;
+        } else {
+          if (tn != type0) {
+            error("Resulting value is of type %s but 'default' is of type %s. "
+                     "Please make sure that both arguments have the same type.", type2char(type0), type2char(tn));
+          }
+          if (!R_compute_identical(PROTECT(getAttrib(outs,R_ClassSymbol)),  PROTECT(getAttrib(na,R_ClassSymbol)), 0)) {
+            error("Resulting value has different class than 'default'. "
+                     "Please make sure that both arguments have the same class.");
+          }
+          UNPROTECT(2);
```

Comment:
do we need to `UNPROTECT(2)` inside the error branch as well?

Reply:
no, we don't

Reply:
I thought we do. This is how we did it in `fifelse`. I will look at the rest of your comments tonight and add what is missing. Thank you

Reply:
we don't need in error branch, but this comment reffers to a aline that is non-error branch, so we do need unprotect here.